### PR TITLE
Add tool for querying events

### PR DESCRIPTION
**Pull request by romac**
Closes: #2869

## Usage

```
USAGE:
    query-events [OPTIONS] --url <URL> <QUERY>

ARGS:
    <QUERY>    The query against which blocks should be matched

OPTIONS:
    -h, --help                   Print help information
        --height <HEIGHT>        The height at which the block should be queried (optional)
        --order <ORDER>          How to order the results by height [default: desc]
        --page <PAGE>            Which page to get [default: 1]
        --per-page <PER_PAGE>    How many results per page [default: 10]
    -u, --url <URL>              The URL of the Tendermint node's RPC endpoint
    -v, --verbose                Enable verbose mode, display partially matching events
```

```
❯ cargo -q run -- -u 'http://provider-sentry-01.goc.earthball.xyz:27001' "send_packet.packet_sequence=890 AND send_packet.packet_src_channel='channel-2' AND send_packet.packet_src_port='provider' AND send_packet.packet_dst_channel='channel-0' AND send_packet.packet_dst_port='consumer'"
2022-11-17T09:17:33.181518Z  INFO query_events: Connecting to http://provider-sentry-01.goc.earthball.xyz:27001/
2022-11-17T09:17:33.182378Z  INFO query_events: Searching for blocks
2022-11-17T09:17:33.690855Z  INFO query_events: Found 2 blocks with partially matching events at heights: 114739, 114467
2022-11-17T09:17:34.183599Z  INFO query_events: Block 114739: found 1 matching events
2022-11-17T09:17:34.183654Z  INFO query_events: [
    Event {
        type_str: "send_packet",
        attributes: [
            Tag {
                key: Key(
                    "packet_data",
                ),
                value: Value(
                    "{\"validator_updates\":[{\"pub_key\":{\"ed25519\":\"5rXiew3ZfIRlu2NCQWABiht1u7jUDR+4CIacHNmMCrs=\"},\"power\":\"27\"}],\"valset_update_id\":\"114739\",\"slash_acks\":[]}",
                ),
            },
            Tag {
                key: Key(
                    "packet_data_hex",
                ),
                value: Value(
                    "7b2276616c696461746f725f75706461746573223a5b7b227075625f6b6579223a7b2265643235353139223a22357258696577335a6649526c75324e43515741426968743175376a5544522b3443496163484e6d4d4372733d227d2c22706f776572223a223237227d5d2c2276616c7365745f7570646174655f6964223a22313134373339222c22736c6173685f61636b73223a5b5d7d",
                ),
            },
            Tag {
                key: Key(
                    "packet_timeout_height",
                ),
                value: Value(
                    "0-0",
                ),
            },
            Tag {
                key: Key(
                    "packet_timeout_timestamp",
                ),
                value: Value(
                    "1670902620888475517",
                ),
            },
            Tag {
                key: Key(
                    "packet_sequence",
                ),
                value: Value(
                    "890",
                ),
            },
            Tag {
                key: Key(
                    "packet_src_port",
                ),
                value: Value(
                    "provider",
                ),
            },
            Tag {
                key: Key(
                    "packet_src_channel",
                ),
                value: Value(
                    "channel-2",
                ),
            },
            Tag {
                key: Key(
                    "packet_dst_port",
                ),
                value: Value(
                    "consumer",
                ),
            },
            Tag {
                key: Key(
                    "packet_dst_channel",
                ),
                value: Value(
                    "channel-0",
                ),
            },
            Tag {
                key: Key(
                    "packet_channel_ordering",
                ),
                value: Value(
                    "ORDER_ORDERED",
                ),
            },
            Tag {
                key: Key(
                    "packet_connection",
                ),
                value: Value(
                    "connection-1",
                ),
            },
        ],
    },
]
2022-11-17T09:17:34.185864Z  INFO query_events: Block 114467: found no matching events
2022-11-17T09:17:34.186081Z  INFO query_events: SUCCESS
```

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by romac** [comment]
@ancazamfir Would you rather be able to specify the max height or the exact height at which to query the block? Or both?

**Comment by ancazamfir** [comment]
> @ancazamfir Would you rather be able to specify the max height or the exact height at which to query the block? Or both?

if specified it should be the exact height. otherwise we can query all as of latest and the user can ignore heights they are not interested in.

**Comment by ancazamfir** [comment]
Maybe it would also be interesting to show the events for the partial match, maybe with a verbose flag.

**Comment by romac** [comment]
@ancazamfir Applied your suggestions, amongst other little improvements.

File: tools/query-events/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    let heights = result
+        .blocks
+        .iter()
+        .map(|b| b.block.header.height)
+        .filter(|h| opts.max_height.map_or(true, |max| h.value() <= max))
+        .collect_vec();
+
+    info!(
+        "Found {} blocks with partially matching events at heights: {}",
+        result.total_count,
```
Comment (ancazamfir) [inline tools/query-events/src/main.rs:86]: When using `--max-height` the count may be heigher, for example with same query as in the description and `--max-height 114467`:
```
2022-11-29T13:41:45.145528Z  INFO query_events: Searching for blocks
2022-11-29T13:41:45.788077Z  INFO query_events: Found 2 blocks with partially matching events at heights: 114467
```
```suggestion
        "Found {} blocks with partially matching events at heights: {}",
        heights.len(),
```

**Review by ancazamfir — APPROVED** [review]
Looks great, thanks @romac. Not sure where to document this tool, together with some typical queries we try when debugging events. Maybe we open an issue to figure it out.

**Pull request merged**